Firestore/core/include/firebase/firestore/timestamp.h

```diff
@@ -0,0 +1,173 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_INCLUDE_FIREBASE_FIRESTORE_TIMESTAMP_H_
+#define FIRESTORE_CORE_INCLUDE_FIREBASE_FIRESTORE_TIMESTAMP_H_
+
+#if !defined(_STLPORT_VERSION)
+#include <chrono>
+#endif
+#include <cstdint>
+#include <ctime>
+
+namespace firebase {
+
+/**
+ * A Timestamp represents a point in time independent of any time zone or
+ * calendar, represented as seconds and fractions of seconds at nanosecond
+ * resolution in UTC Epoch time. It is encoded using the Proleptic Gregorian
+ * Calendar which extends the Gregorian calendar backwards to year one. It is
+ * encoded assuming all minutes are 60 seconds long, i.e. leap seconds are
+ * "smeared" so that no leap second table is needed for interpretation. Range is
+ * from 0001-01-01T00:00:00Z to 9999-12-31T23:59:59.999999999Z.
+ *
+ * @see
+ * https://github.com/google/protobuf/blob/master/src/google/protobuf/timestamp.proto
+ */
+class Timestamp {
+ public:
+  /**
+   * Creates a new timestamp representing the epoch (with seconds and
+   * nanoseconds set to 0).
+   */
+  Timestamp();
+
+  /**
+   * Creates a new timestamp.
+   *
+   * @param seconds The number of seconds of UTC time since Unix epoch
+   *     1970-01-01T00:00:00Z. Must be from 0001-01-01T00:00:00Z to
+   *     9999-12-31T23:59:59Z inclusive.
+   * @param nanoseconds The non-negative fractions of a second at nanosecond
+   *     resolution. Negative second values with fractions must still have
+   *     non-negative nanoseconds values that count forward in time. Must be
+   *     from 0 to 999,999,999 inclusive.
+   */
+  Timestamp(std::int64_t seconds, std::int32_t nanoseconds);
+
+  /**
+   * Creates a new timestamp with the current date.
+   *
+   * If <chrono> is available, precision will correspond to the precision of
+   * `std::chrono::system_clock`. Otherwise, the timestamp will have second
+   * precision.
+   *
+   * @return a new timestamp representing the current date.
+   */
+  static Timestamp Now();
+
+  /**
+   * The number of seconds of UTC time since Unix epoch 1970-01-01T00:00:00Z.
+   */
+  std::int64_t seconds() const {
+    return seconds_;
+  }
+
+  /**
+   * The non-negative fractions of a second at nanosecond resolution. Negative
+   * second values with fractions still have non-negative nanoseconds values
+   * that count forward in time.
+   */
+  std::int32_t nanoseconds() const {
+    return nanoseconds_;
+  }
+
+  /**
+   * Converts `std::time_t` to a Timestamp.
+   *
+   * @param seconds_since_unix_epoch The number of seconds of UTC time since
+   *     Unix epoch 1970-01-01T00:00:00Z. Can be negative to represent dates
+   *     before the epoch.
+   *
+   *     Note that while the epoch of `std::time_t` is unspecified, it's usually
```

Comment:
I'm not sure:
* whether this paragraph break is valid;
* whether we need `<p>` tags.
Can I run a tool to verify the output?

Reply:
I'm not sure either.

You can run `doxygen` on this. (It's apt-get'able, and is also available on homebrew and I'm sure other sources too.) You may have to create a Doxyfile first. I can walk you through the process if you're unfamiliar with the tool.

It also looks like there's some internal tools to do some of this which may be closer to what we'll eventually use, but a bit more work to setup.